File: Client-side/src/app/shared/apis/game.api.ts
/* eslint-disable jsdoc/require-returns-description */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { SportEvent } from 'src/app/shared/interfaces/sport-event.interface';
import { environment } from 'src/environments/environment';
import {
  GameFavoriteList,
  GameList,
  GameListByLabelParams,
  GameListByTypeParams,
  GameListItem,
  GameProviderParams,
  GameSearch,
  GameSort,
  Gamelabel,
  LabelCodeList,
  NewGameList,
  PlayGameParams,
  PlayGameResponse,
  ProviderCategoryInterface,
  ProviderInterface,
  SceneData,
  ScenesGameResponse,
  ScenesInfo,
  ScenesLabelList,
} from '../interfaces/game.interface';
import { ResponseData, ResponseListData } from '../interfaces/response.interface';
import { BaseApi } from './base.api';

@Injectable({
  providedIn: 'root',
})
export class GameApi extends BaseApi {
  /**
   * 获取厂商列表
   *
   * @param category SportsBook, Lottery, Casino, Chess
   * @returns ProviderInterface
   */
  getProviderList(category: string = ''): Observable<ResponseData<ProviderInterface[]>> {
    const url = `${environment.apiUrl}/v1/game/provider`;
    return this.getByCaches(url, { category });
  }

  /**
   * 游戏列表-游戏供应商分类
   *
   * @param param
   * @returns
   */
  gameListByProvider(param: GameProviderParams): Observable<ResponseListData<GameListItem[]>> {
    const url = `${environment.apiUrl}/v1/game/gamelistbyprovider`;
    return this.post(url, param);
  }

  /**
   * 获取游戏类型下拉（交易记录）
   *
   * @returns ProviderCategoryInterface
   */
  getProviderCategory(): Observable<ResponseData<ProviderCategoryInterface[]>> {
    const url = `${environment.apiUrl}/v1/game/providercategory`;
    return this.get(url);
  }

  /**
   * 游戏列表-标签分类 , gameCount 显示游戏数量，预设24个，上限50个
   *
   * @param params GameListByLabelParams
   * @returns
   */
  getGameListByLabel(params: GameListByLabelParams): Observable<ResponseData<GameList>> {
    const url = `${environment.apiUrl}/v1/game/gamelistbylabel`;
    return this.post(url, params);
  }

  /**
   * 新的 为你推荐返回接口 根据标签返回 gameCount 显示游戏数量，预设24个，上限50个
   *
   * @param params GameListByLabelParams
   * @returns
   */
  getGameListByMultipleLabel(params: GameListByLabelParams): Observable<ResponseData<NewGameList>> {
    const url = `${environment.apiUrl}/v1/game/gamemultiplelabel`;
    return this.post(url, params);
  }

  /**
   * 游戏列表-类型分类 , gameCount 显示游戏数量，预设24个，上限50个
   *
   * @param params GameListByTypeParams
   * @returns
   */
  getGamelistbyGameType(params: GameListByTypeParams): Observable<ResponseData<GameList[]>> {
    const url = `${environment.apiUrl}/v1/game/gamelistbygametype`;
    return this.post(url, params);
  }

  /**
   * 获取游戏链接
   *
   * @param params 请求参数
   * @returns
   */
  getPlayGameUrl(params: PlayGameParams): Observable<ResponseData<PlayGameResponse>> {
    const url = `${environment.apiUrl}/v2/game/game/playgame`;
    return this.get<ResponseData<PlayGameResponse>>(url, {
      domain: window.location.origin,
      ...params,
    }).pipe(
      tap(res => {
        if (res?.data?.playGameUrl) this.dataCollectionService.gtmPush('play_game', { game_code: params.gameId });
        if (res?.code === '2096') this.appService.showForbidTip('play');
      }),
    );
  }

  /**
   * 获取游戏排序下拉选项GameSort
   *
   * @returns
   */
  getGameSortList(): Observable<ResponseData<GameSort[]>> {
    const url = `${environment.apiUrl}/v1/game/sortselect`;
    return this.getByCaches(url);
  }

  /**
   * 搜索游戏名称
   *
   * @param name
   * @returns
   */
  getGameByName(name: string): Observable<ResponseData<GameSearch>> {
    const url = `${environment.apiUrl}/v1/game/search`;
    return this.get(url, { name });
  }

  /**
   * 获得游戏详情
   *
   * @param gameId
   * @param providerCatId
   * @returns
   */
  getGameInfo(gameId: string, providerCatId: string): Observable<ResponseData<GameListItem>> {
    const url = `${environment.apiUrl}/v1/game/getgameinfo`;
    return this.get(url, { gameId, providerCatId });
  }

  /**
   * 获取游戏标签下拉
   *
   * @returns
   */
  getGamelabel(): Observable<ResponseData<Gamelabel[]>> {
    const url = `${environment.apiUrl}/v1/game/gamelabel`;
    return this.getByCaches(url);
  }

  /**
   * 根据标签获取厂商列表
   *
   * @param labelCodes
   * @param gameTypes
   * @returns
   */
  getProviderByLabel(labelCodes?: string[], gameTypes?: string[]): Observable<ResponseData<ProviderInterface[]>> {
    const url = `${environment.apiUrl}/v1/game/providerbylabel`;
    return this.post(url, { labelCodes, gameTypes });
  }

  /**
   * 根据标签获取厂商列表
   *
   * @param providerCatId
   * @returns
   */
  getLabelByProviderid(providerCatId: string): Observable<ResponseData<LabelCodeList[]>> {
    const url = `${environment.apiUrl}/v1/game/getlabelbyproviderid`;
    return this.get(url, { providerCatId });
  }

  /**
   * 小游戏收藏列表
   *
   * @param pageIndex
   * @param pageSize
   * @returns
   */
  getFavoritegame(pageIndex: number, pageSize: number): Observable<ResponseData<GameFavoriteList>> {
    const url = `${environment.apiUrl}/v2/game/game/favoritegame`;
    return this.get(url, { pageIndex, pageSize });
  }

  /**
   * 添加收藏游戏
   *
   * @param id
   * @returns
   */
  postAddFavoriteGame(id: string): Observable<ResponseData<{ favoriteCount: number }>> {
    const url = `${environment.apiUrl}/v1/game/addfavoritegame`;
    return this.post(url, { id });
  }

  /**
   * 移除收藏游戏
   *
   * @param id
   * @returns
   */
  postRemoveFavoriteGame(id: string): Observable<ResponseData<{ favoriteCount: number }>> {
    const url = `${environment.apiUrl}/v1/game/removefavoritegame`;
    return this.post(url, { id });
  }

  /**
   * 获取场景标签
   *
   * @returns
   */
  getLabelScenesList(): Observable<ResponseData<ScenesLabelList[]>> {
    const url = `${environment.apiUrl}/v1/game/getlabelsceneslist`;
    return this.get(url);
  }

  /**
   * 获取场景标签-顶部与左侧菜单
   *
   * @returns
   */
  getScenesMenu(): Observable<ResponseData<SceneData>> {
    const url = `${environment.apiUrl}/v2/game/game/getscenesmenu`;
    return this.getByCaches(url);
  }

  /**
   * 通过标签Id获取场景游戏
   *
   * @param ids
   * @returns
   */
  getGameListByIds(ids: number[]): Observable<ResponseData<ScenesGameResponse[]>> {
    const params = '?ids=' + ids.join('&ids=');
    const url = `${environment.apiUrl}/v2/game/game/getscenesgamelistbylabelids${params}`;
    return this.getByCaches<ResponseData<ScenesGameResponse[]>>(url).pipe(
      map(x => {
        if (x?.data) {
          x.data.forEach(item => {
            item.gameList = item.gameList.slice(0, 25);
          });
        }
        return x;
      }),
    );
  }

  /**
   * 最近玩过
   *
   * @param pageIndex
   * @param pageSize
   * @returns
   */
  getRecentLyPlayed(pageIndex: number, pageSize: number): Observable<ResponseListData<GameListItem[]>> {
    const url = `${environment.apiUrl}/v2/game/game/getrecentlyplayed`;
    return this.get(url, { pageIndex, pageSize });
  }

  /**
   * 获取OB熱門賽事
   *
   * @param lang
   * @param matchCount
   * @returns
   */
  getHotMatch(lang: string, matchCount: number): Observable<SportEvent[]> {
    const url = `${environment.apiUrl}/v1/game/gethotmatch`;
    // return this.get<ResponseData<SportEvent[]>>(url, { lang, matchCount }).pipe(
    //   map(v => v?.data || [])
    // );
    return this.get<ResponseData<SportEvent[]>>(url, { lang, matchCount }).pipe(map(v => v?.data || []));
  }

  /**
   * 获取场景标签和游戏列表
   *
   * @param scenesType 场景类型 Hall 大厅列表 Menu 菜单 HallBar 大厅横栏 FrontPage 首页
   * @param gameCount 游戏数量
   * @returns
   */
  getLabelAndGameListByScenes(
    scenesType: 'Hall' | 'Menu' | 'HallBar' | 'FrontPage',
    gameCount: number,
  ): Observable<ResponseData<ScenesInfo[]>> {
    const url = `${environment.apiUrl}/v1/game/getlabelandgamelistbyscenes`;
    return this.post(url, { gameCount, scenesType });
  }
}

File: Flutter-client/lib/common/api/game/game_api.dart
// ignore_for_file: unused_element
import 'package:gogaming_app/config/config.dart';

import '../base/base_api.dart';

enum GameSortType {
  home('HomeSort'),
  provider('ProviderSort'),
  label('LabelSort');

  const GameSortType(this.value);

  final String value;

  factory GameSortType.fromValue(int value) {
    switch (value) {
      case 0:
        return GameSortType.home;
      case 1:
        return GameSortType.provider;
      case 2:
        return GameSortType.label;
    }
    throw 'Invalid enum value: $value';
  }
}

enum GameLabelType {
  hall('Hall'),
  front('FrontPage'),
  menu('Menu'),
  hallBar('HallBar');

  const GameLabelType(this.value);

  final String value;

  factory GameLabelType.fromValue(String value) {
    switch (value) {
      case 'Hall':
        return GameLabelType.hall;
      case 'FrontPage':
        return GameLabelType.front;
      case 'Menu':
        return GameLabelType.menu;
      case 'HallBar':
        return GameLabelType.hallBar;
    }
    throw 'Invalid enum value: $value';
  }
}

enum Game with GoGamingApi {
  /// 游戏信息
  getGameInfo(params: {'providerCatId': '供应商Id-Category', 'gameId': 'gameId'}),

  /// 搜索
  search(params: {'name': "关键词"}),

  /// 获取游戏标签
  gameLabel(),

  /// 获取游戏链接
  playGame(params: {
    'providerCatId': '游戏商代碼',
    'gameId': '游戏id',
    'currencyCode': '用户币别代号',
    'gameCurrencyCode': '游戏币种',
    'playMode': 'Try | Normal',
    'siteType': 'PC | Mobile | APP',
    'domain': '請求方網域名 Web传送站台网域 APP传送系统版本号 ex.ios15.2、Android 11',
    'agOddType': null, //AG盤口類型
  }),

  /// 获取场景标签
  getScenesMenu(),

  /// 根据场景 id 获取游戏列表
  getScenesGameListByLabelIds(params: {
    'ids': [0],
  }),

  /// 移除收藏游戏
  removeFavoriteGame(data: {'id': 0}),

  /// 添加收藏游戏
  addFavoriteGame(data: {'id': 0}),
  getGameListByLabel(data: {
    'labelCodes': <int>[],
    'isRecomment': null, //bool 是否首页推荐
    'gameCount': null, //int 显示游戏数量，预设24个，上限50个
    'isHall': null, // bool 是否游戏大厅
    'entryType': null, // 从大厅进入的类型 0：不区分，1：娱乐城，2：彩票
  }),
  getGameMultipleLabel(data: {
    'labelCodes': <int>[],
    'isRecomment': null, //bool 是否首页推荐
    'gameCount': null, //int 显示游戏数量，预设24个，上限50个
    'isHall': null, // bool 是否游戏大厅
    'entryType': null, // 从大厅进入的类型 0：不区分，1：娱乐城，2：彩票
  }),

  /// * [sortType] 场景类型 `GameSortType.home.value`
  getGameListByProvider(
    data: {
      'labelCode': null,

      /// 变更为 int 数组
      'providerCatIds': null,
      'gameTypes': null,
      'sort': null,
      // HomeSort:0, //首页排序       首页排序，或者不传sortType 默认就是这个
      // ProviderSort:1, //厂商排序    点击厂商的时候传递
      // LabelSort:2, //标签排序    点击标签的时候传递
      'sortType': null,
      'pageIndex': 0,
      'pageSize': 20,
    },
  ),
  getProviderByLabel(
    data: {
      'labelCodes': <String>[],
    },
  ),

  getLabelByProviderId(params: {'providerCatId': 'x'}),
  getSortSelect(),
  getProvider(),

  /// 获取热门游戏
  getHotMatch(params: {
    'matchCount': 0,
    'lang': '',
  }),

  /// 根据场景获取标签和游戏列表
  ///
  /// * [gameCount] 显示游戏数量，预设24个，上限50个
  /// * [scenesType] 场景类型 `GameLabelType.hall.value`
  getLabelAndGameListByScenes(data: {
    'gameCount': 24,
    'scenesType': 'x',
  }),
  getBetRangeSetting(),
  getRecentlyPlayed(
    params: {
      'pageIndex': 0,
      'pageSize': 20,
    },
  ),
  getFavoriteGame(
    params: {
      'pageIndex': 0,
      'pageSize': 20,
    },
  );

  const Game({this.params, this.data});

  @override
  final Map<String, dynamic>? params;
  @override
  final dynamic data;

  @override
  String getBaseUrl() {
    switch (this) {
      case Game.playGame ||
            Game.getRecentlyPlayed ||
            Game.getFavoriteGame ||
            Game.getScenesGameListByLabelIds ||
            Game.getScenesMenu:
        return "${Config.currentConfig.apiUrl}v2";
      default:
        return super.getBaseUrl();
    }
  }

  @override
  String get path {
    switch (this) {
      case Game.getGameMultipleLabel:
        return "/game/gamemultiplelabel";
      case Game.getProvider:
        return "/game/provider";
      case Game.getGameListByLabel:
        return "/game/gamelistbylabel";
      case Game.getLabelAndGameListByScenes:
        return "/game/getlabelandgamelistbyscenes";
      case Game.getRecentlyPlayed:
        return "/game/game/getrecentlyplayed";
      case Game.getFavoriteGame:
        return "/game/game/favoritegame";
      case Game.getGameListByProvider:
        return "/game/gamelistbyprovider";
      case Game.getSortSelect:
        return "/game/sortselect";
      case Game.getProviderByLabel:
        return "/game/providerbylabel";
      case Game.addFavoriteGame:
        return "/game/addfavoritegame";
      case Game.removeFavoriteGame:
        return "/game/removefavoritegame";
      case Game.playGame:
        return '/game/game/playgame';
      case Game.getBetRangeSetting:
        return '/game/getbetrangesetting';
      case Game.search:
        return '/game/search';
      case Game.getGameInfo:
        return '/game/getgameinfo';
      case Game.getLabelByProviderId:
        return '/game/getlabelbyproviderid';
      case Game.gameLabel:
        return '/game/gamelabel';
      case Game.getScenesMenu:
        return '/game/game/getscenesmenu';
      case Game.getScenesGameListByLabelIds:
        return '/game/game/getscenesgamelistbylabelids';
      case Game.getHotMatch:
        return '/game/gethotmatch';
    }
  }

  @override
  HTTPMethod get method {
    switch (this) {
      case Game.getProvider:
      case Game.getSortSelect:
      case Game.playGame:
      case Game.getBetRangeSetting:
      case Game.search:
      case Game.getGameInfo:
      case Game.getLabelByProviderId:
      case Game.gameLabel:
      case Game.getRecentlyPlayed:
      case Game.getFavoriteGame:
      case Game.getScenesGameListByLabelIds:
      case Game.getScenesMenu:
      case Game.getHotMatch:
        return HTTPMethod.get;
      case Game.getGameListByLabel:
      case Game.getGameListByProvider:
      case Game.getProviderByLabel:
      case Game.addFavoriteGame:
      case Game.removeFavoriteGame:
      case Game.getGameMultipleLabel:
      case Game.getLabelAndGameListByScenes:
        return HTTPMethod.post;
    }
  }

  @override
  bool get needCache {
    switch (this) {
      case Game.getBetRangeSetting:
      case Game.getProvider:
      case Game.getSortSelect:
      case Game.getProviderByLabel:
      case Game.getGameListByProvider:
      case Game.getGameMultipleLabel:
      case Game.getGameListByLabel:
      case Game.getLabelAndGameListByScenes:
        return true;
      default:
        return false;
    }
  }
}
